src/solvers/dgmulti/dg.jl

```diff
@@ -485,6 +489,55 @@ function calc_single_boundary_flux!(cache, t, boundary_condition, boundary_key,
   # However, we don't have to re-reshape, since cache.flux_face_values still retains its original shape.
 end
 
+function calc_single_boundary_flux!(cache, t, boundary_condition, boundary_key,
+                                    have_nonconservative_terms::True,
+                                    mesh, equations, dg::DGMulti{NDIMS}) where {NDIMS}
+
+  rd = dg.basis
+  md = mesh.md
+  (; u_face_values, flux_face_values) = cache
+  (; xyzf, nxyzJ, Jf) = md
+  surface_flux, nonconservative_flux = dg.surface_integral.surface_flux
+
+  # reshape face/normal arrays to have size = (num_points_on_face, num_faces_total).
+  # mesh.boundary_faces indexes into the columns of these face-reshaped arrays.
+  num_pts_per_face = rd.Nfq ÷ rd.Nfaces
+  num_faces_total = rd.Nfaces * md.num_elements
+
+  # This function was originally defined as
+  # `reshape_by_face(u) = reshape(view(u, :), num_pts_per_face, num_faces_total)`.
+  # This results in allocations due to https://github.com/JuliaLang/julia/issues/36313.
+  # To avoid allocations, we use Tim Holy's suggestion:
+  # https://github.com/JuliaLang/julia/issues/36313#issuecomment-782336300.
+  reshape_by_face(u) = Base.ReshapedArray(u, (num_pts_per_face, num_faces_total), ())
+
+  u_face_values = reshape_by_face(u_face_values)
+  flux_face_values = reshape_by_face(flux_face_values)
+  Jf = reshape_by_face(Jf)
+  nxyzJ, xyzf = reshape_by_face.(nxyzJ), reshape_by_face.(xyzf) # broadcast over nxyzJ::NTuple{NDIMS,Matrix}
```

Comment:
```suggestion
  u_face_values_reshaped = reshape_by_face(u_face_values)
  flux_face_values_reshaped = reshape_by_face(flux_face_values)
  Jf_reshaped = reshape_by_face(Jf)
  nxyzJ_reshaped, xyzf_reshaped = reshape_by_face.(nxyzJ), reshape_by_face.(xyzf) # broadcast over nxyzJ::NTuple{NDIMS,Matrix}
```
There are a number of type instabilities that can be fixed by assigning new variable names here. Don't ask me why Julia does not recognize this on itself, I just noticed it when running
```julia
@code_warntype Trixi.calc_single_boundary_flux!(semi.cache, t, first(boundary_conditions), first(keys(boundary_conditions)), Trixi.have_nonconservative_terms(equations), mesh, equations, semi.solver)
```
after executing the new elixir once.

Reply:
Thanks! I have some uncommitted changes which address this as well, so I'll commit those and resolve this conversation.